File: android/nestopia/src/main/kotlin/net/sourceforge/nestopia/NestopiaControls.kt
// Copyright (C) 2026 Dominic Opitz
// SPDX-License-Identifier: GPL-2.0-or-later

package net.sourceforge.nestopia

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

@Composable
fun NestopiaControls(
    engine: NestopiaEngine,
    modifier: Modifier = Modifier,
    configuration: NestopiaControllerConfiguration = NestopiaControllerConfiguration(),
) {
    BoxWithConstraints(modifier = modifier) {
        val mode =
            when (configuration.presentationMode) {
                NestopiaControllerPresentationMode.Automatic ->
                    if (maxWidth > maxHeight || maxHeight < 300.dp) {
                        NestopiaControllerPresentationMode.Overlay
                    } else {
                        NestopiaControllerPresentationMode.Gamepad
                    }
                else -> configuration.presentationMode
            }
        val metrics = ControllerMetrics(compact = maxWidth < 500.dp)
        val palette = controllerPalette(configuration)

        if (mode == NestopiaControllerPresentationMode.Overlay) {
            OverlayControls(engine, configuration, metrics, palette)
        } else {
            GamepadControls(engine, configuration, metrics, palette)
        }
    }
}

@Composable
private fun GamepadControls(
    engine: NestopiaEngine,
    configuration: NestopiaControllerConfiguration,
    metrics: ControllerMetrics,
    palette: ControllerPalette,
) {
    val controllerLabel = rememberControllerLabel(configuration)
    val bodyShape =
        when (configuration.theme) {
            NestopiaControllerTheme.System -> RoundedCornerShape(24.dp)
            NestopiaControllerTheme.NES -> RoundedCornerShape(18.dp)
            NestopiaControllerTheme.Famicom -> RoundedCornerShape(12.dp)
        }
    Row(
        modifier =
            Modifier
                .fillMaxSize()
                .padding(horizontal = metrics.outerPadding, vertical = metrics.outerPadding),
        verticalAlignment = Alignment.Bottom,
    ) {
        Row(
            modifier =
                Modifier
                    .fillMaxWidth()
                    .widthIn(max = 680.dp)
                    .shadow(14.dp, bodyShape)
                    .background(bodyPanelBrush(palette.body), bodyShape)
                    .background(bodySheenBrush, bodyShape)
                    .border(1.dp, Color.Black.copy(alpha = 0.22f), bodyShape)
                    .border(1.dp, Color.White.copy(alpha = 0.14f), bodyShape)
                    .padding(metrics.bodyPadding),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.Bottom,
        ) {
            DPad(engine, configuration, metrics, palette, 1f)
            Column(
                modifier =
                    Modifier
                        .padding(horizontal = metrics.sectionSpacing)
                        .background(palette.panel, RoundedCornerShape(9.dp))
                        .padding(horizontal = metrics.utilitySpacing, vertical = metrics.utilitySpacing),
                verticalArrangement = Arrangement.spacedBy(metrics.utilitySpacing),
            ) {
                if (controllerLabel.isNotEmpty()) {
                    Text(
                        text = controllerLabel,
                        color = palette.bodyLabel,
                        fontSize = 10.sp,
                        fontWeight = FontWeight.Black,
                        maxLines = 1,
                    )
                }
                UtilityButtons(engine, configuration, metrics, palette, 1f)
            }
            ActionButtons(engine, configuration, metrics, palette, 1f)
        }
    }
}

@Composable
private fun OverlayControls(
    engine: NestopiaEngine,
    configuration: NestopiaControllerConfiguration,
    metrics: ControllerMetrics,
    palette: ControllerPalette,
) {
    val opacity = configuration.overlayOpacity.coerceIn(0f, 1f)
    Box(
        modifier =
            Modifier
                .fillMaxSize()
                .padding(horizontal = metrics.outerPadding, vertical = metrics.outerPadding),
    ) {
        DPad(
            engine,
            configuration,
            metrics,
            palette,
            opacity,
            Modifier.align(Alignment.BottomStart),
        )
        UtilityButtons(
            engine,
            configuration,
            metrics,
            palette,
            opacity,
            Modifier.align(Alignment.BottomCenter),
        )
        ActionButtons(
            engine,
            configuration,
            metrics,
            palette,
            opacity,
            Modifier.align(Alignment.BottomEnd),
        )
    }
}

// MARK: Directional pad

@Composable
private fun DPad(
    engine: NestopiaEngine,
    configuration: NestopiaControllerConfiguration,
    metrics: ControllerMetrics,
    palette: ControllerPalette,
    opacity: Float,
    modifier: Modifier = Modifier,
) {
    val cell = metrics.direction
    val plateSize = cell * 3
    val density = LocalDensity.current
    val boundsPx =
        remember(plateSize, density) {
            with(density) { Rect(0f, 0f, plateSize.toPx(), plateSize.toPx()) }
        }
    ControllerTouchSurface(
        engine = engine,
        hapticsEnabled = configuration.hapticsEnabled,
        modifier = modifier.size(plateSize),
        resolveActive = { point, _ -> dPadHitTest(point, boundsPx) },
    ) {
        Box(Modifier.size(plateSize)) {
            CrossPlate(size = plateSize, palette = palette, opacity = opacity)
            ArmOverlay(active, NestopiaButton.Up, cell, opacity, Modifier.offset(x = cell, y = 0.dp))
            ArmOverlay(active, NestopiaButton.Left, cell, opacity, Modifier.offset(x = 0.dp, y = cell))
            ArmOverlay(active, NestopiaButton.Right, cell, opacity, Modifier.offset(x = cell * 2, y = cell))
            ArmOverlay(active, NestopiaButton.Down, cell, opacity, Modifier.offset(x = cell, y = cell * 2))
            CenterBoss(size = cell, opacity = opacity, modifier = Modifier.offset(x = cell, y = cell))
        }
    }
}

@Composable
private fun CrossPlate(
    size: Dp,
    palette: ControllerPalette,
    opacity: Float,
) {
    val shape = remember(size) { DPadCrossShape(cornerRadius = size * 0.09f) }
    Box(
        modifier =
            Modifier
                .size(size)
                .shadow(3.dp, shape)
                .background(dPadBrush(palette.directionalPad, opacity), shape)
                .border(1.5.dp, Color.White.copy(alpha = 0.08f * opacity), shape)
                .border(1.dp, Color.Black.copy(alpha = 0.35f * opacity), shape),
    )
}

@Composable
private fun ArmOverlay(
    active: Set<NestopiaButton>,
    button: NestopiaButton,
    cell: Dp,
    opacity: Float,
    modifier: Modifier = Modifier,
) {
    val isActive = active.contains(button)
    val darken by animateOverlayAlpha(isActive, target = 0.24f * opacity)
    Box(
        modifier =
            modifier
                .size(cell)
                .background(Color.Black.copy(alpha = darken)),
    )
}

@Composable
private fun CenterBoss(
    size: Dp,
    opacity: Float,
    modifier: Modifier = Modifier,
) {
    val bossSize = size * 0.6f
    val centerOffset = (size - bossSize) / 2
    Box(
        modifier =
            modifier
                .offset(x = centerOffset, y = centerOffset)
                .size(bossSize)
                .background(
                    Brush.radialGradient(
                        listOf(Color.White.copy(alpha = 0.18f * opacity), Color.Black.copy(alpha = 0.30f * opacity)),
                    ),
                    CircleShape,
                ).border(1.dp, Color.Black.copy(alpha = 0.3f * opacity), CircleShape),
    )
}

// MARK: Utility buttons (SELECT / START)

@Composable
private fun UtilityButtons(
    engine: NestopiaEngine,
    configuration: NestopiaControllerConfiguration,
    metrics: ControllerMetrics,
    palette: ControllerPalette,
    opacity: Float,
    modifier: Modifier = Modifier,
) {
    val density = LocalDensity.current
    val tolerancePx = with(density) { (metrics.utilityWidth * 0.7f).toPx() }
    ControllerTouchSurface(
        engine = engine,
        hapticsEnabled = configuration.hapticsEnabled,
        modifier = modifier,
        resolveActive = { point, frames -> nearestButtonHitTest(point, frames, tolerancePx) },
    ) {
        Row(horizontalArrangement = Arrangement.spacedBy(metrics.utilitySpacing)) {
            UtilityCap(NestopiaButton.Select, "SELECT", metrics, palette, opacity)
            UtilityCap(NestopiaButton.Start, "START", metrics, palette, opacity)
        }
    }
}

@Composable
private fun ControllerTouchScope.UtilityCap(
    button: NestopiaButton,
    label: String,
    metrics: ControllerMetrics,
    palette: ControllerPalette,
    opacity: Float,
) {
    val isActive = active.contains(button)
    val pressedScale by animateOverlayScale(isActive)
    Box(
        modifier =
            Modifier
                .size(metrics.utilityWidth, metrics.utilityHeight)
                .graphicsLayer {
                    scaleX = pressedScale
                    scaleY = pressedScale
                }.background(utilityCapsuleBrush(palette.utilityButtons, opacity), CircleShape)
                .border(1.dp, Color.White.copy(alpha = 0.22f * opacity), CircleShape)
                .semantics { contentDescription = label }
                .reportBounds(button),
        contentAlignment = Alignment.Center,
    ) {
        Text(text = label, color = palette.labels, fontSize = 9.sp, fontWeight = FontWeight.Black)
    }
}

// MARK: Action buttons (A / B)

@Composable
private fun ActionButtons(
    engine: NestopiaEngine,
    configuration: NestopiaControllerConfiguration,
    metrics: ControllerMetrics,
    palette: ControllerPalette,
    opacity: Float,
    modifier: Modifier = Modifier,
) {
    val density = LocalDensity.current
    val tolerancePx = with(density) { (metrics.actionSize * 0.7f).toPx() }
    ControllerTouchSurface(
        engine = engine,
        hapticsEnabled = configuration.hapticsEnabled,
        modifier = modifier,
        resolveActive = { point, frames -> nearestButtonHitTest(point, frames, tolerancePx) },
    ) {
        Row(
            horizontalArrangement = Arrangement.spacedBy(metrics.actionSpacing),
            verticalAlignment = Alignment.Bottom,
        ) {
            Box(Modifier.padding(bottom = metrics.actionSize * 0.22f)) {
                ActionCap(NestopiaButton.B, "B", metrics, palette, opacity)
            }
            ActionCap(NestopiaButton.A, "A", metrics, palette, opacity)
        }
    }
}

@Composable
private fun ControllerTouchScope.ActionCap(
    button: NestopiaButton,
    label: String,
    metrics: ControllerMetrics,
    palette: ControllerPalette,
    opacity: Float,
) {
    val isActive = active.contains(button)
    val pressedScale by animateOverlayScale(isActive)
    val density = LocalDensity.current
    val diameterPx = with(density) { metrics.actionSize.toPx() }
    Column(horizontalAlignment = Alignment.CenterHorizontally) {
        Box(
            modifier =
                Modifier
                    .size(metrics.actionSize)
                    .graphicsLayer {
                        scaleX = pressedScale
                        scaleY = pressedScale
                    }.shadow(3.dp, CircleShape)
                    .background(actionCapBrush(palette.actionButtons, opacity, diameterPx), CircleShape)
                    .border(1.2.dp, Color.White.copy(alpha = 0.32f * opacity), CircleShape)
                    .border(1.dp, Color.Black.copy(alpha = 0.28f * opacity), CircleShape)
                    .semantics { contentDescription = label }
                    .reportBounds(button),
        )
        Text(
            text = label,
            color = palette.bodyLabel.copy(alpha = opacity),
            fontSize = 10.sp,
            fontWeight = FontWeight.Black,
        )
    }
}

@Composable
private fun animateOverlayScale(isActive: Boolean) =
    animateFloatAsState(
        targetValue = if (isActive) 0.93f else 1f,
        animationSpec = tween(durationMillis = 70),
        label = "controllerButtonScale",
    )

@Composable
private fun animateOverlayAlpha(
    isActive: Boolean,
    target: Float,
) = animateFloatAsState(
    targetValue = if (isActive) target else 0f,
    animationSpec = tween(durationMillis = 60),
    label = "controllerArmOverlay",
)

private val bodySheenBrush =
    Brush.verticalGradient(
        listOf(Color.White.copy(alpha = 0.08f), Color.Transparent, Color.Black.copy(alpha = 0.10f)),
    )

private data class ControllerMetrics(
    val direction: Dp,
    val actionSize: Dp,
    val utilityWidth: Dp,
    val utilityHeight: Dp,
    val sectionSpacing: Dp,
    val utilitySpacing: Dp,
    val actionSpacing: Dp,
    val bodyPadding: Dp,
    val outerPadding: Dp,
) {
    constructor(compact: Boolean) : this(
        direction = if (compact) 38.dp else 48.dp,
        actionSize = if (compact) 50.dp else 64.dp,
        utilityWidth = if (compact) 46.dp else 58.dp,
        utilityHeight = if (compact) 26.dp else 32.dp,
        sectionSpacing = if (compact) 4.dp else 18.dp,
        utilitySpacing = if (compact) 4.dp else 10.dp,
        actionSpacing = if (compact) 8.dp else 16.dp,
        bodyPadding = if (compact) 12.dp else 22.dp,
        outerPadding = if (compact) 8.dp else 20.dp,
    )
}

private data class ControllerPalette(
    val body: Color,
    val directionalPad: Color,
    val actionButtons: Color,
    val utilityButtons: Color,
    val labels: Color,
    val bodyLabel: Color,
    val panel: Color,
)

@Composable
private fun controllerPalette(configuration: NestopiaControllerConfiguration): ControllerPalette {
    val defaults =
        when (configuration.theme) {
            NestopiaControllerTheme.System ->
                ControllerPalette(
                    body = MaterialTheme.colorScheme.surfaceContainerHigh.copy(alpha = 0.92f),
                    directionalPad = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.72f),
                    actionButtons = MaterialTheme.colorScheme.primary,
                    utilityButtons = MaterialTheme.colorScheme.secondary,
                    labels = MaterialTheme.colorScheme.onPrimary,
                    bodyLabel = MaterialTheme.colorScheme.onSurface,
                    panel = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.08f),
                )
            NestopiaControllerTheme.NES ->
                ControllerPalette(
                    Color(0xFFB3B3AE),
                    Color(0xFF1A1A1A),
                    Color(0xFFAB0D1F),
                    Color(0xFF292929),
                    Color.White,
                    Color(0xFF262626),
                    Color(0xFF1F1F1F),
                )
            NestopiaControllerTheme.Famicom ->
                ControllerPalette(
                    Color(0xFFE8DBB0),
                    Color(0xFF570A17),
                    Color(0xFFA10814),
                    Color(0xFF6E0D1A),
                    Color(0xFFFFE8B8),
                    Color(0xFF660A17),
                    Color(0xFF570A17),
                )
        }
    val overrides = configuration.colors
    return defaults.copy(
        body = overrides.body ?: defaults.body,
        directionalPad = overrides.directionalPad ?: defaults.directionalPad,
        actionButtons = overrides.actionButtons ?: defaults.actionButtons,
        utilityButtons = overrides.utilityButtons ?: defaults.utilityButtons,
        labels = overrides.labels ?: defaults.labels,
        bodyLabel = overrides.bodyLabel ?: defaults.bodyLabel,
    )
}

File: swift/Sources/Nestopia/NestopiaControls.swift
// Copyright (C) 2026 Dominic Opitz
// SPDX-License-Identifier: GPL-2.0-or-later

import SwiftUI

public struct NestopiaControls: View {
    private let engine: NestopiaEngine
    private let configuration: NestopiaControllerConfiguration

    public init(engine: NestopiaEngine, configuration: NestopiaControllerConfiguration = .init()) {
        self.engine = engine
        self.configuration = configuration
    }

    public var body: some View {
        GeometryReader { proxy in
            let mode = resolvedMode(for: proxy.size)
            let metrics = Metrics(compact: proxy.size.width < 500)
            let palette = Palette(theme: configuration.theme, overrides: configuration.colors)

            Group {
                if mode == .overlay {
                    overlayControls(metrics: metrics, palette: palette)
                } else {
                    gamepadControls(metrics: metrics, palette: palette)
                }
            }
            .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .bottom)
        }
        .accessibilityElement(children: .contain)
        .accessibilityLabel("Game controller")
    }

    private func resolvedMode(for size: CGSize) -> NestopiaControllerPresentationMode {
        guard configuration.presentationMode == .automatic else {
            return configuration.presentationMode
        }
        return size.width > size.height || size.height < 300 ? .overlay : .gamepad
    }

    private func gamepadControls(metrics: Metrics, palette: Palette) -> some View {
        HStack(alignment: .bottom, spacing: 0) {
            dPad(metrics: metrics, palette: palette, opacity: 1)
            Spacer(minLength: metrics.sectionSpacing)
            VStack(spacing: metrics.utilitySpacing) {
                if !configuration.resolvedControllerLabel.isEmpty {
                    Text(configuration.resolvedControllerLabel)
                        .font(.system(size: 10, weight: .black, design: .rounded))
                        .foregroundStyle(palette.bodyLabel)
                        .lineLimit(1)
                        .frame(maxWidth: .infinity, alignment: .leading)
                }
                utilityButtons(metrics: metrics, palette: palette, opacity: 1)
            }
            .padding(.horizontal, metrics.utilitySpacing)
            .padding(.vertical, metrics.utilitySpacing)
            .background(
                RoundedRectangle(cornerRadius: 9, style: .continuous)
                    .fill(palette.panel.opacity(configuration.theme == .system ? 0.12 : 0.94))
            )
            Spacer(minLength: metrics.sectionSpacing)
            actionButtons(metrics: metrics, palette: palette, opacity: 1)
        }
        .padding(metrics.bodyPadding)
        .frame(maxWidth: 680)
        .background { controllerBody(palette: palette) }
        .padding(.horizontal, metrics.outerPadding)
        .padding(.bottom, metrics.outerPadding)
    }

    private func overlayControls(metrics: Metrics, palette: Palette) -> some View {
        ZStack(alignment: .bottom) {
            dPad(metrics: metrics, palette: palette, opacity: configuration.overlayOpacity)
                .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .bottomLeading)
            utilityButtons(
                metrics: metrics, palette: palette, opacity: configuration.overlayOpacity
            )
            .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .bottom)
            actionButtons(metrics: metrics, palette: palette, opacity: configuration.overlayOpacity)
                .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .bottomTrailing)
        }
        .padding(.horizontal, metrics.outerPadding)
        .padding(.bottom, metrics.outerPadding)
    }

    private func controllerBody(palette: Palette) -> some View {
        let shape = controllerBodyShape
        return ZStack {
            if configuration.theme == .system {
                #if compiler(>=6.2)
                    if #available(iOS 26.0, macOS 26.0, tvOS 26.0, *) {
                        shape.fill(.clear)
                            .glassEffect(.regular.tint(palette.body), in: shape)
                    } else {
                        shape.fill(.ultraThinMaterial)
                            .overlay(shape.fill(palette.body))
                    }
                #else
                    shape.fill(.ultraThinMaterial)
                        .overlay(shape.fill(palette.body))
                #endif
            } else {
                shape.fill(bodyPanelGradient(color: palette.body))
                shape.fill(
                    LinearGradient(
                        colors: [.white.opacity(0.08), .clear, .black.opacity(0.10)],
                        startPoint: .top, endPoint: .bottom
                    )
                )
                shape.stroke(.black.opacity(0.22), lineWidth: 1)
            }
            shape.stroke(.white.opacity(configuration.theme == .system ? 0.28 : 0.12), lineWidth: 1)
        }
        .shadow(color: .black.opacity(0.24), radius: 14, y: 8)
    }

    private var controllerBodyShape: AnyShape {
        switch configuration.theme {
        case .system: AnyShape(RoundedRectangle(cornerRadius: 24, style: .continuous))
        case .nes: AnyShape(RoundedRectangle(cornerRadius: 18, style: .continuous))
        case .famicom: AnyShape(RoundedRectangle(cornerRadius: 12, style: .continuous))
        }
    }

    // MARK: - Directional pad

    #if os(tvOS)
        private func dPad(metrics: Metrics, palette: Palette, opacity: Double) -> some View {
            let cell = metrics.direction
            let bounds = CGSize(width: cell * 3, height: cell * 3)
            return ZStack {
                crossPlate(size: bounds, palette: palette, opacity: opacity)
                Grid(horizontalSpacing: 0, verticalSpacing: 0) {
                    GridRow {
                        Color.clear.frame(width: cell, height: cell)
                        tvDirectionButton(.up, "▲", cell: cell, palette: palette, opacity: opacity)
                        Color.clear.frame(width: cell, height: cell)
                    }
                    GridRow {
                        tvDirectionButton(
                            .left, "◀", cell: cell, palette: palette, opacity: opacity)
                        centerBoss(size: cell, opacity: opacity)
                        tvDirectionButton(
                            .right, "▶", cell: cell, palette: palette, opacity: opacity)
                    }
                    GridRow {
                        Color.clear.frame(width: cell, height: cell)
                        tvDirectionButton(
                            .down, "▼", cell: cell, palette: palette, opacity: opacity)
                        Color.clear.frame(width: cell, height: cell)
                    }
                }
            }
            .frame(width: bounds.width, height: bounds.height)
        }

        private func tvDirectionButton(
            _ button: NestopiaControllerButton, _ glyph: String, cell: CGFloat, palette: Palette,
            opacity: Double
        ) -> some View {
            Button {
                engine.setButton(button, pressed: true)
                DispatchQueue.main.asyncAfter(deadline: .now() + 0.05) {
                    engine.setButton(button, pressed: false)
                }
            } label: {
                Text(glyph)
                    .font(.system(size: 18, weight: .black, design: .rounded))
                    .foregroundStyle(palette.labels)
                    .frame(width: cell, height: cell)
                    .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
        }
    #else
        private func dPad(metrics: Metrics, palette: Palette, opacity: Double) -> some View {
            let cell = metrics.direction
            let bounds = CGSize(width: cell * 3, height: cell * 3)
            return ControllerTouchSurface(
                engine: engine,
                hapticsEnabled: configuration.hapticsEnabled,
                resolveActive: { point, _ in
                    dPadHitTest(at: point, in: CGRect(origin: .zero, size: bounds))
                },
                content: { active in
                    ZStack {
                        crossPlate(size: bounds, palette: palette, opacity: opacity)
                        Grid(horizontalSpacing: 0, verticalSpacing: 0) {
                            GridRow {
                                Color.clear.frame(width: cell, height: cell)
                                armOverlay(
                                    .up, active: active, cell: cell, opacity: opacity)
                                Color.clear.frame(width: cell, height: cell)
                            }
                            GridRow {
                                armOverlay(
                                    .left, active: active, cell: cell, opacity: opacity)
                                centerBoss(size: cell, opacity: opacity)
                                armOverlay(
                                    .right, active: active, cell: cell, opacity: opacity)
                            }
                            GridRow {
                                Color.clear.frame(width: cell, height: cell)
                                armOverlay(
                                    .down, active: active, cell: cell, opacity: opacity)
                                Color.clear.frame(width: cell, height: cell)
                            }
                        }
                    }
                    .frame(width: bounds.width, height: bounds.height)
                }
            )
        }
    #endif

    private func crossPlate(size: CGSize, palette: Palette, opacity: Double) -> some View {
        let shape = DPadCrossShape(cornerRadius: size.width * 0.09)
        return ZStack {
            shape.fill(dPadGradient(color: palette.dPad, opacity: opacity))
            shape.stroke(.white.opacity(0.08 * opacity), lineWidth: 1.5)
            shape.stroke(.black.opacity(0.35 * opacity), lineWidth: 1)
        }
        .frame(width: size.width, height: size.height)
        .shadow(color: .black.opacity(0.2 * opacity), radius: 3, y: 2)
    }

    private func armOverlay(
        _ button: NestopiaControllerButton, active: Set<NestopiaControllerButton>, cell: CGFloat,
        opacity: Double
    ) -> some View {
        Rectangle()
            .fill(.black.opacity(active.contains(button) ? 0.24 * opacity : 0))
            .frame(width: cell, height: cell)
            .animation(.easeOut(duration: 0.06), value: active.contains(button))
    }

    private func centerBoss(size: CGFloat, opacity: Double) -> some View {
        Circle()
            .fill(
                RadialGradient(
                    colors: [.white.opacity(0.18 * opacity), .black.opacity(0.30 * opacity)],
                    center: .center, startRadius: 0, endRadius: size * 0.32
                )
            )
            .overlay(Circle().stroke(.black.opacity(0.3 * opacity), lineWidth: 1))
            .frame(width: size * 0.6, height: size * 0.6)
    }

    // MARK: - Utility buttons (SELECT / START)

    #if os(tvOS)
        private func utilityButtons(
            metrics: Metrics, palette: Palette, opacity: Double
        ) -> some View {
            HStack(spacing: metrics.utilitySpacing) {
                tvUtilityButton(
                    .select, "SELECT", metrics: metrics, palette: palette, opacity: opacity)
                tvUtilityButton(
                    .start, "START", metrics: metrics, palette: palette, opacity: opacity)
            }
        }

        private func tvUtilityButton(
            _ button: NestopiaControllerButton, _ label: String, metrics: Metrics, palette: Palette,
            opacity: Double
        ) -> some View {
            Button {
                engine.setButton(button, pressed: true)
                DispatchQueue.main.asyncAfter(deadline: .now() + 0.05) {
                    engine.setButton(button, pressed: false)
                }
            } label: {
                utilityCapVisual(
                    label: label, size: metrics.utility, palette: palette, opacity: opacity,
                    isActive: false)
            }
            .buttonStyle(.plain)
        }
    #else
        private func utilityButtons(
            metrics: Metrics, palette: Palette, opacity: Double
        ) -> some View {
            ControllerTouchSurface(
                engine: engine,
                hapticsEnabled: configuration.hapticsEnabled,
                resolveActive: { point, frames in
                    nearestButtonHitTest(
                        at: point, among: frames, tolerance: metrics.utility.width * 0.7)
                },
                content: { active in
                    HStack(spacing: metrics.utilitySpacing) {
                        utilityCap(
                            .select, label: "SELECT", metrics: metrics, palette: palette,
                            opacity: opacity, active: active)
                        utilityCap(
                            .start, label: "START", metrics: metrics, palette: palette,
                            opacity: opacity, active: active)
                    }
                }
            )
        }

        private func utilityCap(
            _ button: NestopiaControllerButton, label: String, metrics: Metrics, palette: Palette,
            opacity: Double, active: Set<NestopiaControllerButton>
        ) -> some View {
            utilityCapVisual(
                label: label, size: metrics.utility, palette: palette, opacity: opacity,
                isActive: active.contains(button)
            )
            .reportingControllerFrame(of: button)
        }
    #endif

    private func utilityCapVisual(
        label: String, size: CGSize, palette: Palette, opacity: Double, isActive: Bool
    ) -> some View {
        Capsule()
            .fill(utilityCapsuleGradient(color: palette.utility, opacity: opacity))
            .overlay(Capsule().strokeBorder(.white.opacity(0.22 * opacity), lineWidth: 1))
            .overlay(
                Text(label)
                    .font(.system(size: 9, weight: .black, design: .rounded))
                    .foregroundStyle(palette.labels)
            )
            .frame(width: size.width, height: size.height)
            .scaleEffect(isActive ? 0.94 : 1)
            .brightness(isActive ? -0.12 : 0)
            .contentShape(Capsule())
            .animation(.easeOut(duration: 0.07), value: isActive)
            .accessibilityLabel(label)
    }

    // MARK: - Action buttons (A / B)

    #if os(tvOS)
        private func actionButtons(
            metrics: Metrics, palette: Palette, opacity: Double
        ) -> some View {
            HStack(alignment: .bottom, spacing: metrics.actionSpacing) {
                tvActionButton(.b, "B", metrics: metrics, palette: palette, opacity: opacity)
                    .padding(.bottom, metrics.action.height * 0.22)
                tvActionButton(.a, "A", metrics: metrics, palette: palette, opacity: opacity)
            }
        }

        private func tvActionButton(
            _ button: NestopiaControllerButton, _ label: String, metrics: Metrics, palette: Palette,
            opacity: Double
        ) -> some View {
            Button {
                engine.setButton(button, pressed: true)
                DispatchQueue.main.asyncAfter(deadline: .now() + 0.05) {
                    engine.setButton(button, pressed: false)
                }
            } label: {
                actionCapVisual(
                    label: label, diameter: metrics.action.width, palette: palette,
                    opacity: opacity, isActive: false)
            }
            .buttonStyle(.plain)
        }
    #else
        private func actionButtons(
            metrics: Metrics, palette: Palette, opacity: Double
        ) -> some View {
            ControllerTouchSurface(
                engine: engine,
                hapticsEnabled: configuration.hapticsEnabled,
                resolveActive: { point, frames in
                    nearestButtonHitTest(
                        at: point, among: frames, tolerance: metrics.action.width * 0.7)
                },
                content: { active in
                    HStack(alignment: .bottom, spacing: metrics.actionSpacing) {
                        actionCap(
                            .b, label: "B", metrics: metrics, palette: palette,
                            opacity: opacity, active: active
                        )
                        .padding(.bottom, metrics.action.height * 0.22)
                        actionCap(
                            .a, label: "A", metrics: metrics, palette: palette,
                            opacity: opacity, active: active)
                    }
                }
            )
        }

        private func actionCap(
            _ button: NestopiaControllerButton, label: String, metrics: Metrics, palette: Palette,
            opacity: Double, active: Set<NestopiaControllerButton>
        ) -> some View {
            actionCapVisual(
                label: label, diameter: metrics.action.width, palette: palette, opacity: opacity,
                isActive: active.contains(button)
            )
            .reportingControllerFrame(of: button)
        }
    #endif

    private func actionCapVisual(
        label: String, diameter: CGFloat, palette: Palette, opacity: Double, isActive: Bool
    ) -> some View {
        VStack(spacing: 4) {
            Circle()
                .fill(
                    actionCapGradient(color: palette.action, opacity: opacity, diameter: diameter)
                )
                .overlay(Circle().strokeBorder(.white.opacity(0.32 * opacity), lineWidth: 1.2))
                .overlay(Circle().stroke(.black.opacity(0.28 * opacity), lineWidth: 1))
                .frame(width: diameter, height: diameter)
                .scaleEffect(isActive ? 0.93 : 1)
                .brightness(isActive ? -0.12 : 0)
                .shadow(color: .black.opacity(0.28 * opacity), radius: 3, y: 2)
                .contentShape(Circle())
            Text(label)
                .font(.system(size: 10, weight: .black, design: .rounded))
                .foregroundStyle(palette.bodyLabel)
                .opacity(opacity)
        }
        .animation(.easeOut(duration: 0.07), value: isActive)
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(label)
    }
}

private struct Metrics {
    let direction: CGFloat
    let action: CGSize
    let utility: CGSize
    let sectionSpacing: CGFloat
    let utilitySpacing: CGFloat
    let actionSpacing: CGFloat
    let bodyPadding: CGFloat
    let outerPadding: CGFloat

    init(compact: Bool) {
        direction = compact ? 38 : 48
        action = CGSize(width: compact ? 50 : 64, height: compact ? 50 : 64)
        utility = CGSize(width: compact ? 46 : 58, height: compact ? 26 : 32)
        sectionSpacing = compact ? 4 : 18
        utilitySpacing = compact ? 4 : 10
        actionSpacing = compact ? 8 : 16
        bodyPadding = compact ? 12 : 22
        outerPadding = compact ? 8 : 20
    }
}

private struct Palette {
    let body: Color
    let dPad: Color
    let action: Color
    let utility: Color
    let labels: Color
    let bodyLabel: Color
    let panel: Color

    init(theme: NestopiaControllerTheme, overrides: NestopiaControllerColorOverrides) {
        let defaults: (Color, Color, Color, Color, Color, Color, Color)
        switch theme {
        case .system:
            defaults = (
                .primary.opacity(0.08), .primary.opacity(0.68), .accentColor,
                .primary.opacity(0.55), .white, .primary,
                .primary.opacity(0.12)
            )
        case .nes:
            defaults = (
                Color(red: 0.70, green: 0.70, blue: 0.68), Color(white: 0.10),
                Color(red: 0.67, green: 0.05, blue: 0.12), Color(white: 0.16), .white,
                Color(white: 0.15),
                Color(white: 0.12)
            )
        case .famicom:
            defaults = (
                Color(red: 0.91, green: 0.86, blue: 0.69),
                Color(red: 0.34, green: 0.04, blue: 0.09),
                Color(red: 0.63, green: 0.03, blue: 0.08),
                Color(red: 0.43, green: 0.05, blue: 0.10),
                Color(red: 0.98, green: 0.91, blue: 0.72),
                Color(red: 0.40, green: 0.04, blue: 0.09),
                Color(red: 0.34, green: 0.04, blue: 0.09)
            )
        }
        body = overrides.body ?? defaults.0
        dPad = overrides.directionalPad ?? defaults.1
        action = overrides.actionButtons ?? defaults.2
        utility = overrides.utilityButtons ?? defaults.3
        labels = overrides.labels ?? defaults.4
        bodyLabel = overrides.bodyLabel ?? defaults.5
        panel = defaults.6
    }
}
